### PR TITLE
nautilus: mgr/fs/volumes misc fixes

### PR DESCRIPTION
**Pull request by ajarr**
Fixes: https://tracker.ceph.com/issues/46235
Fixes: https://tracker.ceph.com/issues/46478
Fixes: https://tracker.ceph.com/issues/46466

### CODE REVIEW

**Comment by smithfarm** [comment]
```
Python Unit-test Failure

FAIL: test_pool_create (tasks.mgr.dashboard.test_pool.PoolTest)

2020-07-18 13:27:50,157.157 INFO:__main__:----------------------------------------------------------------------
2020-07-18 13:27:50,157.157 INFO:__main__:Traceback (most recent call last):
2020-07-18 13:27:50,157.157 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/dashboard/test_pool.py", line 267, in test_pool_create
2020-07-18 13:27:50,157.157 INFO:__main__:    self.assertIn(conf, new_pool['configuration'])
2020-07-18 13:27:50,157.157 INFO:__main__:AssertionError: {'name': 'rbd_qos_bps_limit', 'source': 1, 'value': '2048'} not found in []
2020-07-18 13:27:50,157.157 INFO:__main__:
2020-07-18 13:27:50,157.157 INFO:__main__:----------------------------------------------------------------------
2020-07-18 13:27:50,157.157 INFO:__main__:Ran 82 tests in 1001.796s
2020-07-18 13:27:50,158.158 INFO:__main__:
2020-07-18 13:27:50,158.158 INFO:__main__:
```

**Comment by smithfarm** [comment]
jenkins test dashboard backend

**Comment by smithfarm** [comment]
@ajarr There are a couple more that you might consider including here:

* https://tracker.ceph.com/issues/46388
* https://tracker.ceph.com/issues/46464

**Comment by smithfarm** [comment]
```
2020-07-18 16:45:02,758.758 INFO:__main__:----------------------------------------------------------------------
2020-07-18 16:45:02,759.759 INFO:__main__:Traceback (most recent call last):
2020-07-18 16:45:02,759.759 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/mgr/test_module_selftest.py", line 320, in test_selftest_cluster_log
2020-07-18 16:45:02,759.759 INFO:__main__:    priority, message)
2020-07-18 16:45:02,759.759 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-dashboard-pr-backend/qa/tasks/ceph_test_case.py", line 97, in __exit__
2020-07-18 16:45:02,760.760 INFO:__main__:    raise AssertionError("Expected log message not found: '{0}'".format(expected_pattern))
2020-07-18 16:45:02,760.760 INFO:__main__:AssertionError: Expected log message not found: '[WRN] foo bar warning'
2020-07-18 16:45:02,760.760 INFO:__main__:
2020-07-18 16:45:02,760.760 INFO:__main__:----------------------------------------------------------------------
2020-07-18 16:45:02,761.761 INFO:__main__:Ran 197 tests in 3474.817s
2020-07-18 16:45:02,761.761 INFO:__main__:
2020-07-18 16:45:02,761.761 INFO:__main__:
```

**Comment by ajarr** [comment]
> @ajarr There are a couple more that you might consider including here:

Thanks, @smithfarm. The fixes for the following two issues are in different parts of the code base, so I preferred to handle them as separate PRs.

> 
> * https://tracker.ceph.com/issues/46388

Even though the conflicts are minor, I've asked the PR author to backport it.


> * https://tracker.ceph.com/issues/46464

The PR is here https://github.com/ceph/ceph/pull/36180

**Comment by smithfarm** [comment]
jenkins test dashboard backend

**Comment by ajarr** [comment]
http://pulpito.front.sepia.ceph.com/rraja-2020-07-20_06:51:46-fs-ajarr-nautilus-testing-2020-07-18-distro-basic-smithi/

**Comment by ajarr** [comment]
@kotreshhr can you quickly take a look at this PR?

**Comment by ajarr** [comment]
New round of testing after fixing the `import typing` error

http://pulpito.front.sepia.ceph.com/rraja-2020-07-20_12:31:26-fs-ajarr-nautilus-testing-2020-07-20-distro-basic-smithi/

**Comment by ajarr** [comment]
http://pulpito.front.sepia.ceph.com/rraja-2020-07-21_13:12:54-fs-ajarr-nautilus-testing-2020-07-21-distro-basic-smithi/

File: src/pybind/mgr/volumes/fs/volume.py [hunk 1]
```diff
@@ -98,9 +98,24 @@ def delete_fs_volume(self, volname, confirm):
                 "that is what you want, re-issue the command followed by " \
                 "--yes-i-really-mean-it.".format(volname)
 
+        ret, out, err = self.mgr.check_mon_command({
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py:101]: @kotreshhr tests are failing now with,
```
AttributeError: 'Module' object has no attribute 'check_mon_command'
```

`check_mon_command` is not in nautilus.  Is it OK just to use `mon_command` instead?

Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/volume.py:101]: Yes, mon_command can be used. The only difference between them is that 'check_mon_command' raised if ret !=0

Comment (ajarr) [reply src/pybind/mgr/volumes/fs/volume.py]: As discussed that'd raise an unhandled exception. If ret !=0 I just return the output of mon_command().

Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/volume.py]: lgtm

**Comment by ajarr** [comment]
http://pulpito.front.sepia.ceph.com/rraja-2020-07-22_15:24:27-fs-ajarr-nautilus-testing-2020-07-22-distro-basic-smithi/

**Comment by ajarr** [comment]
I now see that all tests fail in test_volume_create() when trying to remove the volume created by the test.

I see this in the mgr log 
```
2020-07-22 18:55:00.182 7f9ea47dd700 10 mgr.server _allowed_command  client.admin capable
2020-07-22 18:55:00.182 7f9ea47dd700  0 log_channel(audit) log [DBG] : from='client.18473 -' entity='client.admin' cmd=[{"yes-i-really-mean-it": "--yes-i-really-mean-it", "prefix": "fs volume rm", "vol_name": "volume_906", "target": ["mgr", ""]}]: dispatch
2020-07-22 18:55:00.182 7f9ea47dd700 10 mgr.server _handle_command passing through 4
2020-07-22 18:55:00.182 7f9ea4fde700 20 mgr Gil Switched to new thread state 0x56015b9b10d0
2020-07-22 18:55:00.182 7f9ea4fde700  1 -- 172.21.15.169:0/12022 --> [v2:172.21.15.169:3300/0,v1:172.21.15.169:6789/0] -- mon_command({"prefix": "config get", "who": "mon", "key": "mon_allow_pool_delete", "format": "json"} v 0) v1 -- 0x56015fd65400 con 0x560158f68d80
2020-07-22 18:55:00.182 7f9eb56c5700  1 -- 172.21.15.169:0/12022 <== mon.1 v2:172.21.15.169:3300/0 11406 ==== mon_command_ack([{"prefix": "config get", "who": "mon", "key": "mon_allow_pool_delete", "format": "json"}]=-22 unrecognized entity 'mon' v146) v1 ==== 147+0+0 (crc 0 0 0) 0x5601589bf180 con 0x560158f68d80
2020-07-22 18:55:00.182 7f9ea2fda700  1 -- 172.21.15.169:0/12022 --> [v2:172.21.15.169:3300/0,v1:172.21.15.169:6789/0] -- mon_get_version(what=osdmap handle=4) v1 -- 0x560159c9a000 con 0x560158f68d80
2020-07-22 18:55:00.182 7f9eb56c5700  1 -- 172.21.15.169:0/12022 <== mon.1 v2:172.21.15.169:3300/0 11407 ==== mon_get_version_reply(handle=4 version=711) v2 ==== 24+0+0 (crc 0 0 0) 0x5601581f44c0 con 0x560158f68d80
2020-07-22 18:55:00.182 7f9eb3ec2700 10 MonCommandCompletion::finish()
2020-07-22 18:55:00.182 7f9eb3ec2700 20 mgr Gil Switched to new thread state 0x56015b3bc580
2020-07-22 18:55:00.182 7f9eb3ec2700 20 mgr ~Gil Destroying new thread state 0x56015b3bc580
2020-07-22 18:55:00.182 7f9eb3ec2700 10 mgr notify_all notify_all: notify_all command
2020-07-22 18:55:00.182 7f9ea4fde700 20 mgr[volumes] mon_command: 'config get' -> -22 in 0.001s
2020-07-22 18:55:00.182 7f9ea4fde700 20 mgr ~Gil Destroying new thread state 0x56015b9b10d0
2020-07-22 18:55:00.182 7f9ea4fde700 -1 mgr.server reply reply (22) Invalid argument unrecognized entity 'mon'
2020-07-22 18:55:00.182 7f9ea4fde700  1 -- [v2:172.21.15.169:6800/12022,v1:172.21.15.169:6801/12022] --> 172.21.15.161:0/2540246774 -- command_reply(tid 0: -22 unrecognized entity 'mon') v1 -- 0x56015b90a960 con 0x56016061c480
```

File: src/pybind/mgr/volumes/fs/volume.py [hunk 2]
```diff
@@ -98,12 +98,14 @@ def delete_fs_volume(self, volname, confirm):
                 "that is what you want, re-issue the command followed by " \
                 "--yes-i-really-mean-it.".format(volname)
 
-        ret, out, err = self.mgr.check_mon_command({
+        ret, out, err = self.mgr.mon_command({
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/volume.py:101]: from the test failure looks like this is incorrect usage of the method.

Comment (kotreshhr) [reply src/pybind/mgr/volumes/fs/volume.py:101]: I tested it locally. The config command is not taking 'mon', it is expecting specific mon, like mon.0. 
It works with 'mon.*' though, so please change it and it should work.

File: src/pybind/mgr/volumes/fs/volume.py [hunk 3]
```diff
@@ -98,12 +98,14 @@ def delete_fs_volume(self, volname, confirm):
...
             'prefix': 'config get',
             'key': 'mon_allow_pool_delete',
             'who': 'mon',
```
Comment (kotreshhr) [inline src/pybind/mgr/volumes/fs/volume.py:104]: s/'who' : 'mon'/'who' : 'mon.*'/

Comment (ajarr) [reply src/pybind/mgr/volumes/fs/volume.py:104]: Done. Thanks!

**Comment by ajarr** [comment]
http://pulpito.front.sepia.ceph.com/rraja-2020-07-24_11:33:52-fs-ajarr-nautilus-testing-2020-07-24-distro-basic-smithi/

The failures are unrelated. This is good to merge.

**Comment by ajarr** [comment]
I see this error now,
 ```
020-07-24T14:10:24.831 INFO:tasks.cephfs_test_runner:======================================================================
2020-07-24T14:10:24.832 INFO:tasks.cephfs_test_runner:ERROR: test_volume_rm_arbitrary_pool_removal (tasks.cephfs.test_volumes.TestVolumes)
2020-07-24T14:10:24.832 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-07-24T14:10:24.832 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-07-24T14:10:24.833 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_ceph_ceph-c_ajarr-nautilus-testing-2020-07-24/qa/tasks/cephfs/test_volumes.py", line 325, in test_volume_rm_arbitrary_pool_removal
2020-07-24T14:10:24.833 INFO:tasks.cephfs_test_runner:    vol_status = json.loads(self._fs_cmd("status", self.volname, "--format=json-pretty"))
2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:    return _default_decoder.decode(s)
2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python3.6/json/decoder.py", line 339, in decode
2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
2020-07-24T14:10:24.835 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python3.6/json/decoder.py", line 357, in raw_decode
2020-07-24T14:10:24.835 INFO:tasks.cephfs_test_runner:    raise JSONDecodeError("Expecting value", s, err.value) from None
2020-07-24T14:10:24.835 INFO:tasks.cephfs_test_runner:json.decoder.JSONDecodeError: Expecting value: line 2 column 1 (char 1)
2020-07-24T14:10:24.836 INFO:tasks.cephfs_test_runner:
```

**Comment by kotreshhr** [comment]
> I see this error now here,
> http://pulpito.front.sepia.ceph.com/rraja-2020-07-24_11:33:52-fs-ajarr-nautilus-testing-2020-07-24-distro-basic-smithi/5253495/teuthology.log
> 
> ```
> 020-07-24T14:10:24.831 INFO:tasks.cephfs_test_runner:======================================================================
> 2020-07-24T14:10:24.832 INFO:tasks.cephfs_test_runner:ERROR: test_volume_rm_arbitrary_pool_removal (tasks.cephfs.test_volumes.TestVolumes)
> 2020-07-24T14:10:24.832 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2020-07-24T14:10:24.832 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2020-07-24T14:10:24.833 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/github.com_ceph_ceph-c_ajarr-nautilus-testing-2020-07-24/qa/tasks/cephfs/test_volumes.py", line 325, in test_volume_rm_arbitrary_pool_removal
> 2020-07-24T14:10:24.833 INFO:tasks.cephfs_test_runner:    vol_status = json.loads(self._fs_cmd("status", self.volname, "--format=json-pretty"))
> 2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python3.6/json/__init__.py", line 354, in loads
> 2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:    return _default_decoder.decode(s)
> 2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python3.6/json/decoder.py", line 339, in decode
> 2020-07-24T14:10:24.834 INFO:tasks.cephfs_test_runner:    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
> 2020-07-24T14:10:24.835 INFO:tasks.cephfs_test_runner:  File "/usr/lib/python3.6/json/decoder.py", line 357, in raw_decode
> 2020-07-24T14:10:24.835 INFO:tasks.cephfs_test_runner:    raise JSONDecodeError("Expecting value", s, err.value) from None
> 2020-07-24T14:10:24.835 INFO:tasks.cephfs_test_runner:json.decoder.JSONDecodeError: Expecting value: line 2 column 1 (char 1)
> 2020-07-24T14:10:24.836 INFO:tasks.cephfs_test_runner:
> ```
> 

This is because 'ceph fs status' cmd's json support patch is missing. I think this needs following two patches.

1.  4da6381fb1e
2. 138117f2f48

> And this earlier in teuthology.log,
> 
> ```
> 2020-07-24T13:32:19.990 INFO:tasks.ceph.mgr.y.smithi154.stderr:Exception in thread puregejob.2:
> 2020-07-24T13:32:19.990 INFO:tasks.ceph.mgr.y.smithi154.stderr:Traceback (most recent call last):
> 2020-07-24T13:32:19.991 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/lib64/python2.7/threading.py", line 812, in __bootstrap_inner
> 2020-07-24T13:32:19.991 INFO:tasks.ceph.mgr.y.smithi154.stderr:    self.run()
> 2020-07-24T13:32:19.991 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/share/ceph/mgr/volumes/fs/async_job.py", line 61, in run
> 2020-07-24T13:32:19.991 INFO:tasks.ceph.mgr.y.smithi154.stderr:    self.async_job.unregister_async_job(vol_job[0], vol_job[1], thread_id)
> 2020-07-24T13:32:19.991 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/share/ceph/mgr/volumes/fs/async_job.py", line 152, in unregister_async_job
> 2020-07-24T13:32:19.992 INFO:tasks.ceph.mgr.y.smithi154.stderr:    log.debug("unregistering async job {0}.{1} from thread {2}".format(volname, job, thread_id))
> 2020-07-24T13:32:19.992 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/lib64/python2.7/logging/__init__.py", line 1137, in debug
> 2020-07-24T13:32:19.992 INFO:tasks.ceph.mgr.y.smithi154.stderr:    self._log(DEBUG, msg, args, **kwargs)
> 2020-07-24T13:32:19.992 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/lib64/python2.7/logging/__init__.py", line 1268, in _log
> 2020-07-24T13:32:19.992 INFO:tasks.ceph.mgr.y.smithi154.stderr:    self.handle(record)
> 2020-07-24T13:32:19.993 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/lib64/python2.7/logging/__init__.py", line 1278, in handle
> 2020-07-24T13:32:19.993 INFO:tasks.ceph.mgr.y.smithi154.stderr:    self.callHandlers(record)
> 2020-07-24T13:32:19.993 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/lib64/python2.7/logging/__init__.py", line 1318, in callHandlers
> 2020-07-24T13:32:19.993 INFO:tasks.ceph.mgr.y.smithi154.stderr:    hdlr.handle(record)
> 2020-07-24T13:32:19.993 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/lib64/python2.7/logging/__init__.py", line 749, in handle
> 2020-07-24T13:32:19.994 INFO:tasks.ceph.mgr.y.smithi154.stderr:    self.emit(record)
> 2020-07-24T13:32:19.994 INFO:tasks.ceph.mgr.y.smithi154.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 65, in emit
> 2020-07-24T13:32:19.994 INFO:tasks.ceph.mgr.y.smithi154.stderr:    self._module._ceph_log(ceph_level, self.format(record))
> 2020-07-24T13:32:19.994 INFO:tasks.ceph.mgr.y.smithi154.stderr:TypeError: must be string without null bytes, not str
> 2020-07-24T13:32:19.994 INFO:tasks.ceph.mgr.y.smithi154.stderr:
> ```

Not sure about this.

**Comment by kotreshhr** [comment]
I verified that both tests pass locally after pulling in 4da6381 and 138117f

```

2020-07-26 23:13:44,886.886 INFO:__main__:test_volume_rm_when_mon_delete_pool_false (tasks.cephfs.test_volumes.TestVolumes) ... ok
2020-07-26 23:13:44,887.887 INFO:__main__:Stopped test: test_volume_rm_when_mon_delete_pool_false (tasks.cephfs.test_volumes.TestVolumes) in 21.876021s
2020-07-26 23:13:44,887.887 INFO:__main__:
2020-07-26 23:13:44,888.888 INFO:__main__:----------------------------------------------------------------------
2020-07-26 23:13:44,888.888 INFO:__main__:Ran 1 test in 21.877s
2020-07-26 23:13:44,888.888 INFO:__main__:
2020-07-26 23:13:44,888.888 INFO:__main__:OK


2020-07-26 23:14:47,771.771 INFO:__main__:test_volume_rm_arbitrary_pool_removal (tasks.cephfs.test_volumes.TestVolumes) ... ok
2020-07-26 23:14:47,772.772 INFO:__main__:Stopped test: test_volume_rm_arbitrary_pool_removal (tasks.cephfs.test_volumes.TestVolumes) in 18.754503s
2020-07-26 23:14:47,772.772 INFO:__main__:
2020-07-26 23:14:47,773.773 INFO:__main__:----------------------------------------------------------------------
2020-07-26 23:14:47,773.773 INFO:__main__:Ran 1 test in 18.755s
2020-07-26 23:14:47,773.773 INFO:__main__:
2020-07-26 23:14:47,773.773 INFO:__main__:OK

```

**Comment by ajarr** [comment]
> I verified that both tests pass locally after pulling in [4da6381](https://github.com/ceph/ceph/commit/4da6381fb1ec90a5280ca54b33ee9527993482c1) and [138117f](https://github.com/ceph/ceph/commit/138117f2f48c4cca014334b4547b229c1f05807e)

Thanks, Kotresh. But testing [4da6381](https://github.com/ceph/ceph/commit/4da6381fb1ec90a5280ca54b33ee9527993482c1) would require running rados/mgr suite. Can we avoid that?

**Comment by kotreshhr** [comment]
> > I verified that both tests pass locally after pulling in [4da6381](https://github.com/ceph/ceph/commit/4da6381fb1ec90a5280ca54b33ee9527993482c1) and [138117f](https://github.com/ceph/ceph/commit/138117f2f48c4cca014334b4547b229c1f05807e)
> 
> Thanks, Kotresh. But testing [4da6381](https://github.com/ceph/ceph/commit/4da6381fb1ec90a5280ca54b33ee9527993482c1) would require running rados/mgr suite. Can we avoid that?

Yes, we can workaround by using the 'ceph osd poo ls' with 'detail' option to validate whether any pool has volume name as metadata. I have sent you the patch. Please apply it and run the test. Hopefully, there are no further surprises this time :)

**Comment by ajarr** [comment]
@kotreshhr do we still need https://github.com/ceph/ceph/commit/138117f2f48c4cca014334b4547b229c1f05807e ?

**Comment by kotreshhr** [comment]
> @kotreshhr do we still need [138117f](https://github.com/ceph/ceph/commit/138117f2f48c4cca014334b4547b229c1f05807e) ?

Nope, we don't need for this now.

**Comment by ajarr** [comment]
http://pulpito.front.sepia.ceph.com/rraja-2020-07-27_15:19:52-fs-ajarr-nautilus-testing-2020-07-27-distro-basic-smithi/

**Comment by ajarr** [comment]
failures were unrelated.

**Comment by neha-ojha** [comment]
Looks like https://github.com/ceph/ceph/pull/33325 is follow-on fix for https://github.com/ceph/ceph/pull/33194/commits/87292811215f6ded9a784d3216a910faaef648e2 based on https://github.com/ceph/ceph/pull/33325#issue-375370206. Was it backported to nautilus with https://github.com/ceph/ceph/pull/36167/commits/9f323e0acb6f84fe827af69b0bf366b772e4db15?

**Comment by batrick** [comment]
No, @ajarr please create a followup backport for 8729281.

**Comment by yuriw** [comment]
@ajarr @batrick @neha-ojha https://tracker.ceph.com/issues/46763

**Comment by yuriw** [comment]
see https://github.com/ceph/ceph/pull/36377

**Comment by ajarr** [comment]
@neha-ojha @batrick @yuriw thanks! missed that it was modifying ceph_test_case and needed rados suite testing.

**Pull request merged**